RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cu

```diff
@@ -0,0 +1,498 @@
+}
+
+void CAHitQuadrupletGeneratorGPU::allocateOnGPU() {
+  cudaStreamCreateWithFlags(&cudaStream_, cudaStreamNonBlocking);
+
+  cudaMallocHost(&h_doublets, maxNumberOfLayerPairs * sizeof(GPULayerDoublets));
+
+  cudaMallocHost(&h_indices,
+                 maxNumberOfLayerPairs * maxNumberOfDoublets * 2 * sizeof(int));
+  cudaMallocHost(&h_x, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMallocHost(&h_y, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMallocHost(&h_z, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMallocHost(&h_rootLayerPairs, maxNumberOfRootLayerPairs * sizeof(int));
+
+  cudaMalloc(&d_indices,
+             maxNumberOfLayerPairs * maxNumberOfDoublets * 2 * sizeof(int));
+  cudaMalloc(&d_doublets, maxNumberOfLayerPairs * sizeof(GPULayerDoublets));
+  cudaMalloc(&d_layers, maxNumberOfLayers * sizeof(GPULayerHits));
+  cudaMalloc(&d_x, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMalloc(&d_y, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMalloc(&d_z, maxNumberOfLayers * maxNumberOfHits * sizeof(float));
+  cudaMalloc(&d_rootLayerPairs,
+             maxNumberOfRootLayerPairs * sizeof(unsigned int));
+  //////////////////////////////////////////////////////////
+  // ALLOCATIONS FOR THE INTERMEDIATE RESULTS (STAYS ON WORKER)
+  //////////////////////////////////////////////////////////
+
+  cudaMalloc(&device_theCells,
+             maxNumberOfLayerPairs * maxNumberOfDoublets * sizeof(GPUCACell));
+
+  cudaMalloc(&device_isOuterHitOfCell,
+             maxNumberOfLayers * maxNumberOfHits *
+                 sizeof(GPUSimpleVector<maxCellsPerHit, unsigned int>));
+  cudaMemset(device_isOuterHitOfCell, 0,
+             maxNumberOfLayers * maxNumberOfHits *
+                 sizeof(GPUSimpleVector<maxCellsPerHit, unsigned int>));
+  h_foundNtuplets.resize(maxNumberOfRegions);
+  d_foundNtuplets.resize(maxNumberOfRegions);
+  for (int i = 0; i < maxNumberOfRegions; ++i) {
+    cudaMalloc(&d_foundNtuplets[i],
+               sizeof(GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>));
+    cudaMallocHost(&h_foundNtuplets[i],
+                   sizeof(GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>));
+  }
+
+  cudaMallocHost(&tmp_layers, maxNumberOfLayers * sizeof(GPULayerHits));
+  cudaMallocHost(&tmp_layerDoublets,
+                 maxNumberOfLayerPairs * sizeof(GPULayerDoublets));
+  cudaMallocHost(&h_layers, maxNumberOfLayers * sizeof(GPULayerHits));
+}
+
+void CAHitQuadrupletGeneratorGPU::launchKernels(const TrackingRegion &region,
+                                                int regionIndex) {
+
+  assert(regionIndex < maxNumberOfRegions);
+  dim3 numberOfBlocks_create(32, numberOfLayerPairs);
+  dim3 numberOfBlocks_connect(16, numberOfLayerPairs);
+  dim3 numberOfBlocks_find(8, numberOfRootLayerPairs);
+  ((GPUSimpleVector<maxNumberOfQuadruplets, Quadruplet>
+        *)(h_foundNtuplets[regionIndex]))
+      ->reset();
```

Comment:
If `h_foundNtuplets` really has to contain `void *`, could we at least use `reinterpret_cast` here (and elsewhere)?